test/e2e/test/elasticsearch/checks_es.go

```diff
@@ -185,29 +171,173 @@ func (e *esClusterChecks) CheckESNodesTopology() test.Step {
 	}
 }
 
-func compareRoles(expected *esv1.Node, actualRoles []string) bool {
+func (e *esClusterChecks) compareTopology(es esv1.Elasticsearch, topoElem esv1.NodeSet, node client.Node, nodeStats client.NodeStats) error {
+	// get config to check roles
+	v, err := version.Parse(es.Spec.Version)
+	if err != nil {
+		return err
+	}
+	cfg := esv1.DefaultCfg(v)
+	if err := esv1.UnpackConfig(topoElem.Config, v, &cfg); err != nil {
+		return err
+	}
+	if err = compareRoles(cfg.Node, node.Roles); err != nil {
+		return err
+	}
+
+	// ES returns a string, parse it as an int64, base10
+	if err = compareCgroupMemoryLimit(topoElem, nodeStats); err != nil {
+		return err
+	}
+
+	if err = compareCgroupCPULimit(topoElem, nodeStats.OS.CGroup.CPU); err != nil {
+		return err
+	}
+
+	// get pods to check ressources requirements
+	pods, err := e.k.GetPods(test.ESPodListOptionsByNodeSet(e.Elasticsearch.Namespace, e.Elasticsearch.Name, topoElem.Name)...)
+	if err != nil {
+		return err
+	}
+	if len(pods) == 0 {
+		return fmt.Errorf("no pod found for ES %q / nodeSet %q", e.Elasticsearch.Name, topoElem.Name)
+	}
+	if err = compareSpecResources(topoElem, pods); err != nil {
+		return err
+	}
+
+	return compareClaimedStorage(e.k, topoElem, pods)
+}
+
+func compareRoles(expected *esv1.Node, actualRoles []string) error {
 	for _, r := range actualRoles {
 		if !expected.HasRole(esv1.NodeRole(r)) {
-			return false
+			return fmt.Errorf("actual role %s not expected", r)
 		}
 	}
-
-	return true
+	return nil
 }
 
-func compareMemoryLimit(topologyElement esv1.NodeSet, cgroupMemoryLimitsInBytes int64) bool {
+// compareCgroupMemoryLimit compares the memory limit specified in a nodeSet with the limit set in the CPU control group at the OS level
+func compareCgroupMemoryLimit(topologyElement esv1.NodeSet, nodeStats client.NodeStats) error {
 	var memoryLimit *resource.Quantity
 	for _, c := range topologyElement.PodTemplate.Spec.Containers {
 		if c.Name == esv1.ElasticsearchContainerName {
 			memoryLimit = c.Resources.Limits.Memory()
 		}
 	}
-	if memoryLimit == nil {
+	if memoryLimit == nil || memoryLimit.IsZero() {
 		// no expected memory, consider it's ok
-		return true
+		return nil
+	}
+
+	actualCgroupMemoryLimit, err := strconv.ParseInt(
+		nodeStats.OS.CGroup.Memory.LimitInBytes, 10, 64,
+	)
+	if err != nil {
+		return err
+	}
+	expectedCgroupMemoryLimit := memoryLimit.Value()
+	if expectedCgroupMemoryLimit != actualCgroupMemoryLimit {
+		return fmt.Errorf("expected Cgroup memory limit %d, got %d", expectedCgroupMemoryLimit, actualCgroupMemoryLimit)
+	}
+	return nil
+}
+
+// compareCgroupCPULimit compares the CPU limit specified in a nodeSet with the limit set in the CPU control group at the OS level
+func compareCgroupCPULimit(topologyElement esv1.NodeSet, cgroupCPU struct {
+	CFSPeriodMicros int `json:"cfs_period_micros"`
+	CFSQuotaMicros  int `json:"cfs_quota_micros"`
```

Comment:
I don't think the tags are required here. I guess I understand why you don't want to use `nodeStats client.NodeStats` (as we do for memory), however my feeling is that it would make it easier to read:

```golang
func compareCgroupCPULimit(topologyElement esv1.NodeSet, nodeStats client.NodeStats) error {
	var expectedCPULimit *resource.Quantity
	for _, c := range topologyElement.PodTemplate.Spec.Containers {
		if c.Name == esv1.ElasticsearchContainerName {
			expectedCPULimit = c.Resources.Limits.Cpu()
		}
	}
	if expectedCPULimit == nil || expectedCPULimit.IsZero() {
		// no expected cpu, consider it's ok
		return nil
	}

	cgroupCPU := nodeStats.OS.CGroup.CPU
	actualCgroupCPULimit := float64(cgroupCPU.CFSQuotaMicros) / float64(cgroupCPU.CFSPeriodMicros)
	if expectedCPULimit.AsApproximateFloat64() != actualCgroupCPULimit {
		return fmt.Errorf("expected Cgroup CPU limit [%f], got [%f]", expectedCPULimit.AsApproximateFloat64(), actualCgroupCPULimit)
	}
	return nil
}
```